MapboxCoreNavigation/Offline.swift

```diff
@@ -0,0 +1,122 @@
+import Foundation
+import MapboxDirections
+import MapboxNavigationNative
+
+public enum RequestOption {
+    case preferServerSide // TODO: Not yet implemented
+    case preferClientSide // TODO: Not yet implemented
+    case serverSideOnly
+    case clientSideOnly
+}
+
+public enum OfflineRoutingError: Error {
+    case unexpectedRouteResult
+    case corruptRouteData
+}
+
+class OfflineStorage: NSObject {
+    
+    static var tilesPath: URL {
+        let path = coreNavigationCacheDirectory.appendingPathComponent("tiles")
+        ensureDirectoryExist(at: path)
+        return path
+    }
+    
+    static var translationsPath: URL {
+        let path = coreNavigationCacheDirectory.appendingPathComponent("translations")
+        ensureDirectoryExist(at: path)
+        return path
+    }
+    
+    static var coreNavigationCacheDirectory: URL {
+        let cacheDirectory = NSSearchPathForDirectoriesInDomains(.cachesDirectory, .userDomainMask, true)[0]
+        return URL(fileURLWithPath: cacheDirectory).appendingPathComponent(Bundle.mapboxCoreNavigation.bundleIdentifier!).appendingPathComponent("offline")
+    }
+    
+    // Ensures that a directory exist at the given path.
+    // Returns false if no directory was created.
+    @discardableResult
+    static func ensureDirectoryExist(at path: URL) -> Bool {
+        do {
+            try FileManager.default.createDirectory(at: path, withIntermediateDirectories: true, attributes: nil)
+            return true
+        } catch {
+            return false
+        }
+    }
+}
+
+public typealias RouteCompletionHandler = (_ waypoints: [Waypoint]?, _ routes: [Route]?, _ error: NSError?) -> Void
+
+public class ExtendedDirections: NSObject {
```

Comment:
`OfflineDirections` might be a better name. Perhaps the functionality could be added as an extension on the existing `Directions` object.

Reply:
What if `OfflineDirections` and `Directions` both conformed to a `DirectionsProtocol` protocol (although we'd still need to figure out naming).